runtime/integration-tests/src/generic/envs/runtime_env.rs

```diff
@@ -46,6 +55,258 @@ impl<T: Runtime> Default for RuntimeEnv<T> {
+		&mut self,
+		caller: Keyring,
+		value: U256,
+		contract: impl Into<String>,
+		function: impl Into<String>,
+		args: Option<&[Token]>,
+	) -> Result<CallInfo, DispatchError> {
+		self.call(caller, value, contract, function, args)
+	}
+
+	fn call_mut(
+		&mut self,
+		caller: Keyring,
+		value: U256,
+		contract: impl Into<String>,
+		function: impl Into<String>,
+		args: Option<&[Token]>,
+	) -> Result<CallInfo, DispatchError> {
+		self.call_mut(caller, value, contract, function, args)
+	}
+
+	fn view(
+		&mut self,
+		caller: Keyring,
+		contract: impl Into<String>,
+		function: impl Into<String>,
+		args: Option<&[Token]>,
+	) -> Result<CallInfo, DispatchError> {
+		self.call(caller, U256::zero(), contract, function, args)
+	}
+}
+
+impl<T: Runtime> RuntimeEnv<T> {
+	pub fn call_mut(
+		&mut self,
+		caller: Keyring,
+		value: U256,
+		contract: impl Into<String>,
+		function: impl Into<String>,
+		args: Option<&[Token]>,
+	) -> Result<CallInfo, DispatchError> {
+		let contract_info = self
+			.deployed_contracts
+			.get(&contract.into())
+			.expect("Contract not deployed")
+			.clone();
+		let input = contract_info
+			.contract
+			.functions_by_name(function.into().as_ref())
+			.expect(ESSENTIAL)
+			.iter()
+			.filter_map(|f| f.encode_input(args.unwrap_or_default()).ok())
+			.collect::<Vec<_>>()
+			.pop()
+			.expect("No matching function Signature found.");
+
+		self.parachain_state_mut(|| {
+			let (base_fee, _) = <T as pallet_evm::Config>::FeeCalculator::min_gas_price();
+
+			<T as pallet_evm::Config>::Runner::call(
+				caller.into(),
+				sp_core::H160::from(contract_info.address().0),
+				input,
+				value,
+				GAS_LIMIT,
+				Some(base_fee),
+				None,
+				None,
+				Vec::new(),
+				// NOTE: Taken from pallet-evm implementation
+				VALIDATE,
+				// NOTE: Taken from pallet-evm implementation
+				TRANSACTIONAL,
+				None,
+				None,
+				<T as pallet_evm::Config>::config(),
+			)
+			.map_err(|re| re.error.into())
+		})
+	}
+
+	pub fn call(
+		// TODO: Needs to imutable actually, but the current state implementation does
+		//       not rollback but error out upon changes, which is not ideal if you want to
+		//       test stuff without altering your state.
+		&mut self,
+		caller: Keyring,
+		value: U256,
+		contract: impl Into<String>,
+		function: impl Into<String>,
+		args: Option<&[Token]>,
+	) -> Result<CallInfo, DispatchError> {
+		let contract_info = self
+			.deployed_contracts
+			.get(&contract.into())
+			.expect("Contract not deployed")
+			.clone();
+		let input = contract_info
+			.contract
+			.function(function.into().as_ref())
+			.expect(ESSENTIAL)
+			.encode_input(args.unwrap_or_default())
+			.expect(ESSENTIAL);
+
+		// TODO: Needs to imutable actually, but the current state implementation does
+		//       not rollback but error out upon changes, which is not ideal if you want
+		// to       test stuff without altering your state.
+		self.parachain_state_mut(|| {
```

Comment:
Interesting. Maybe we should change the `*_state()` behavior to make a rollback instead

Reply:
We should definitely. There should be state changes to be allowed in `parachain_state` that get just rolled back.

Reply:
Where can I add that?

Reply:
We can wrap the `f()` call inside `RuntimeEnv::parachain_state()` with a transaction and revert in case of an error. For reference: https://github.com/centrifuge/centrifuge-chain/blob/main/pallets/loans/src/lib.rs#L635